CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc

```diff
@@ -117,7 +149,11 @@ class SiPixelLorentzAnglePCLWorker : public DQMGlobalEDAnalyzer<SiPixelLorentzAn
   double clustChargeMax_;
   int hist_depth_;
   int hist_drift_;
-
+    
+  std::unique_ptr<TFile> hFile_;
+  std::shared_ptr<TTree> SiPixelLorentzAngleTreeBarrel_;
```

Comment:
Does these two TTrees have to be std::shared_ptr?

Reply:
I'm not sure. The shared_ptr allows for multiple pointers so I chose it. Is there other recommendation?

Reply:
I think std::unique_ptr would also work but should be checked. Some sort of general recommendation is not to use std::shared_ptr unless really necessary.

Reply:
@wweiphy unless these pointers need to be shared in other classes, which is not the case here, `std::unique_ptr` is recommended

Reply:
I actually tested std::unique_ptr before but the program crashed due to segmentation violation. 
When I use shared_ptr it works well.

Reply:
@wweiphy I cannot produce a segfault by using this code:
```diff
diff --git a/CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc b/CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc
index 07483273325..25abeeec4d5 100644
--- a/CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc
+++ b/CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLWorker.cc
@@ -148,8 +148,8 @@ private:
   int hist_drift_;
 
   std::unique_ptr<TFile> hFile_;
-  std::shared_ptr<TTree> SiPixelLorentzAngleTreeBarrel_;
-  std::shared_ptr<TTree> SiPixelLorentzAngleTreeForward_;
+  std::unique_ptr<TTree> SiPixelLorentzAngleTreeBarrel_;
+  std::unique_ptr<TTree> SiPixelLorentzAngleTreeForward_;
 
   // es consumes
   edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> geomEsToken_;
@@ -255,7 +255,7 @@ SiPixelLorentzAnglePCLWorker::SiPixelLorentzAnglePCLWorker(const edm::ParameterS
   //    Barrel pixel
   hFile_ = std::make_unique<TFile>(filename_.c_str(), "RECREATE");
   SiPixelLorentzAngleTreeBarrel_ =
-      std::make_shared<TTree>("SiPixelLorentzAngleTreeBarrel_", "SiPixel LorentzAngle tree barrel", bufsize);
+      std::make_unique<TTree>("SiPixelLorentzAngleTreeBarrel_", "SiPixel LorentzAngle tree barrel", bufsize);
   SiPixelLorentzAngleTreeBarrel_->Branch("run", &run_, "run/I", bufsize);
   SiPixelLorentzAngleTreeBarrel_->Branch("event", &event_, "event/l", bufsize);
   SiPixelLorentzAngleTreeBarrel_->Branch("lumiblock", &lumiblock_, "lumiblock/I", bufsize);
@@ -292,7 +292,7 @@ SiPixelLorentzAnglePCLWorker::SiPixelLorentzAnglePCLWorker(const edm::ParameterS
   //    Forward pixel
 
   SiPixelLorentzAngleTreeForward_ =
-      std::make_shared<TTree>("SiPixelLorentzAngleTreeForward_", "SiPixel LorentzAngle tree forward", bufsize);
+      std::make_unique<TTree>("SiPixelLorentzAngleTreeForward_", "SiPixel LorentzAngle tree forward", bufsize);
   SiPixelLorentzAngleTreeForward_->Branch("run", &run_, "run/I", bufsize);
   SiPixelLorentzAngleTreeForward_->Branch("event", &event_, "event/l", bufsize);
   SiPixelLorentzAngleTreeForward_->Branch("lumiblock", &lumiblock_, "lumiblock/I", bufsize);
```

runnning:
```
cmsRun test.py outName=testrun.root maxEvents=10 inputFiles=/store/data/Run2018D/SingleMuon/ALCARECO/SiPixelCalSingleMuon-ForPixelALCARECO_UL2018-v1/30000/6F0401BA-B0B9-8F46-B2FC-732EE66154F0.root
```

can you check again?

Reply:
@wweiphy can you check again the point above?
Then I think we can merge this.

Reply:
@mmusich I still got segmentation violation when I used std::unique_ptr 

```
Thread 2 (Thread 0x7f131c30d700 (LWP 15660)):
#0  0x00007f133a6171d9 in waitpid () from /lib64/libpthread.so.0
#1  0x00007f1334656387 in edm::service::cmssw_stacktrace_fork() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#2  0x00007f133465701a in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#3  0x00007f133ac0faf0 in std::execute_native_thread_routine (__p=0x7f131c98e5f0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#4  0x00007f133a60fea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f133a3379fd in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f1338493540 (LWP 15259)):
#0  0x00007f133a32cccd in poll () from /lib64/libc.so.6
#1  0x00007f13346567b7 in full_read.constprop () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#2  0x00007f13346570ec in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#3  0x00007f133465a59b in sig_dostack_then_abort () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x00007f12efb9b45f in SiPixelLorentzAnglePCLWorker::~SiPixelLorentzAnglePCLWorker() () from /uscms_data/d3/wwei/LA/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/pluginCalibTrackerSiPixelLorentzAngle.so
#6  0x00007f12efb9b5cf in SiPixelLorentzAnglePCLWorker::~SiPixelLorentzAnglePCLWorker() () from /uscms_data/d3/wwei/LA/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/pluginCalibTrackerSiPixelLorentzAngle.so
#7  0x00007f133353b8b3 in std::_Sp_counted_deleter<edm::global::EDProducerBase*, std::default_delete<edm::global::EDProducerBase>, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/pluginIOMCRandomEnginePlugins.so
#8  0x00007f133353c4e2 in std::_Sp_counted_ptr_inplace<edm::maker::ModuleHolderT<edm::global::EDProducerBase>, std::allocator<edm::maker::ModuleHolderT<edm::global::EDProducerBase> >, (__gnu_cxx::_Lock_policy)2>::_M_dispose() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/pluginIOMCRandomEnginePlugins.so
#9  0x00007f133cdadc8c in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, edm::propagate_const<std::shared_ptr<edm::maker::ModuleHolder> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, edm::propagate_const<std::shared_ptr<edm::maker::ModuleHolder> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, edm::propagate_const<std::shared_ptr<edm::maker::ModuleHolder> > > > >::_M_erase(std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, edm::propagate_const<std::shared_ptr<edm::maker::ModuleHolder> > > >*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#10 0x00007f133cdadc16 in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, edm::propagate_const<std::shared_ptr<edm::maker::ModuleHolder> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, edm::propagate_const<std::shared_ptr<edm::maker::ModuleHolder> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, edm::propagate_const<std::shared_ptr<edm::maker::ModuleHolder> > > > >::_M_erase(std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, edm::propagate_const<std::shared_ptr<edm::maker::ModuleHolder> > > >*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#11 0x00007f133cdadce6 in std::_Sp_counted_ptr<edm::ModuleRegistry*, (__gnu_cxx::_Lock_policy)2>::_M_dispose() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#12 0x00007f133ccc4e08 in std::default_delete<edm::Schedule>::operator()(edm::Schedule*) const [clone .isra.0] [clone .part.0] () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#13 0x00007f133cccc1c2 in edm::EventProcessor::~EventProcessor() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#14 0x000000000040b851 in (anonymous namespace)::EventProcessorWithSentry::~EventProcessorWithSentry() ()
#15 0x000000000040b649 in main ()

Current Modules:

Module: none (crashed)

A fatal system signal has occurred: segmentation violation
Segmentation fault
```

Reply:
what configuration file are you running?
I can't reproduce :/

Reply:
Here it is:
https://github.com/wweiphy/LA/blob/main/test.py

Reply:
> Here it is:
> https://github.com/wweiphy/LA/blob/main/test.py

I think this is private, or at least when I click to the link, it leads to a 404 error

Reply:
Sorry, it should be visible now.
https://github.com/wweiphy/LA/blob/main/test.py

Reply:
@wweiphy, 
I continue not being able to reproduce your problem.
Here is my full recipe:
```
cmsrel CMSSW_12_0_X_2021-07-14-2300
cd CMSSW_12_0_X_2021-07-14-2300/src/
cmsenv
git cms-addpkg CalibTracker/SiPixelLorentzAngle
git cms-addpkg Calibration/TkAlCaRecoProducers/
git cms-addpkg Configuration/AlCa
git cms-addpkg Configuration/EventContent
git cms-addpkg Configuration/PyReleaseValidation
git cms-addpkg Configuration/StandardSequences/
git remote add wweiphy git@github.com:wweiphy/cmssw.git
git fetch wweiphy
git checkout wweiphy/LA-PCL
git apply git.patch  # take the patch from https://github.com/CMSTrackerDPG/cmssw/pull/15#discussion_r661543887
scramv1 b -j 20
wget https://raw.githubusercontent.com/wweiphy/LA/main/test.py
cmsRun test.py
cmsRun test.py outName=testrun.root maxEvents=10 inputFiles=/store/data/Run2018D/SingleMuon/ALCARECO/SiPixelCalSingleMuon-ForPixelALCARECO_UL2018-v1/30000/6F0401BA-B0B9-8F46-B2FC-732EE66154F0.root
```

both `cmsRun` instances work just fine for me.

Reply:
@mmusich I tried again in AFS and I got 

```
Fatal system signal has occurred during exit
Aborted (core dumped)
```

Reply:
And still, when I use std::shared_ptr I did not get the fatal information.

Reply:
Please just make it `std::unique`

Reply:
I cannot reproduce the error either. I think we should just go ahead with the unique pointer.